### ISSUE TITLE
Panic on addOutputsToLocalScope

### ISSUE THREAD

**Issue by duongnt** [argoproj/argo#3505]
Checklist:

 * [x] I've included the version.
 * [ ] I've included reproduction steps.
 * [ ] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**: 
I'm not sure if stable is also non-viable or is it just me, but I'm getting
```
runtime error: invalid memory address or nil pointer dereference" stack="goroutine 343 [running]:
runtime/debug.Stack(0xc05fda1f51, 0x173d180, 0x28720d0)
  /usr/local/go/src/runtime/debug/stack.go:24 +0x9d
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).operate.func2(0xc0023d40b0)
  /go/src/github.com/argoproj/argo/workflow/controller/operator.go:164 +0xc4
panic(0x173d180, 0x28720d0)
  /usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).addOutputsToLocalScope(0xc0023d40b0, 0xc002a460a0, 0x15, 0xc0024fa540, 0xc002071a90)
  /go/src/github.com/argoproj/argo/workflow/controller/operator.go:2072 +0x493
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).buildLocalScope(0xc0023d40b0, 0xc002071a90, 0xc002a460a0, 0x15, 0xc0023ddb80)
 /go/src/github.com/argoproj/argo/workflow/controller/operator.go:2050 +0xaa
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).buildLocalScopeFromTask(0xc0023d40b0, 0xc0018ddea0, 0xc001001ce0, 0x0, 0x0, 0x0)
  /go/src/github.com/argoproj/argo/workflow/controller/dag.go:475 +0x24c
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).resolveDependencyReferences(0xc0023d40b0, 0xc0018ddea0, 0xc001001ce0, 0xf, 0x0, 0x0)
  /go/src/github.com/argoproj/argo/workflow/controller/dag.go:485 +0x5d
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc0023d40b0, 0xc0018ddea0, 0xc003465870, 0xf)
  /go/src/github.com/argoproj/argo/workflow/controller/dag.go:372 +0x7e6
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).executeDAG(0xc0023d40b0, 0xc002520160, 0x1f, 0xc0024faa00, 0xc0005167b0, 0x25, 0xc0048778c0, 0x1bba5a0, 0xc0023d4160, 0xc00243def0, ...)
  /go/src/github.com/argoproj/argo/workflow/controller/dag.go:231 +0x2d6
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).executeTemplate(0xc0023d40b0, 0xc002520160, 0x1f, 0x1bba5a0, 0xc0023d4160, 0xc0024fa980, 0x0, 0x0, 0x0, 0x0, ...)
  /go/src/github.com/argoproj/argo/workflow/controller/operator.go:1517 +0x1de4
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).operate(0xc0023d40b0)
  /go/src/github.com/argoproj/argo/workflow/controller/operator.go:290 +0x125d
github.com/argoproj/argo/workflow/controller.(*WorkflowController).processNextItem(0xc000097c00, 0x0)
 /go/src/github.com/argoproj/argo/workflow/controller/controller.go:433 +0x8c2
github.com/argoproj/argo/workflow/controller.(*WorkflowController).runWorker(0xc000097c00)
 /go/src/github.com/argoproj/argo/workflow/controller/controller.go:359 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc002e4f020)
  /go/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002e4f020, 0x3b9aca00, 0x0, 0x1, 0xc0000abc80)
  /go/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc002e4f020, 0x3b9aca00, 0xc0000abc80)
  /go/pkg/mod/k8s.io/apimachinery@v0.16.7-beta.0/pkg/util/wait/wait.go:88 +0x4d
created by github.com/argoproj/argo/workflow/controller.(*WorkflowController).Run
 /go/src/github.com/argoproj/argo/workflow/controller/controller.go:172 +0x975
```
with the stable build.

Not sure if there's any change I need to make in my workflow definition.

**What you expected to happen**: no panic

**How to reproduce it (as minimally and precisely as possible)**: 

**Anything else we need to know?**: I upgraded from 2.8.0 to. stable and all workflows that involve output passing failed with this panic.

**Environment**:

- Argo version: 
```
$ argo version
argo: v2.8.0
  BuildDate: 2020-05-11T22:55:43Z
  GitCommit: 8f696174746ed01b9bf1941ad03da62d312df641
  GitTreeState: clean
  GitTag: v2.8.0
  GoVersion: go1.13.4
  Compiler: gc
  Platform: darwin/amd64
```
- Kubernetes version :
```
$ kubectl version -o yaml
clientVersion:
  buildDate: "2020-02-13T18:08:14Z"
  compiler: gc
  gitCommit: 06ad960bfd03b39c8310aaf92d1e7c12ce618213
  gitTreeState: clean
  gitVersion: v1.17.3
  goVersion: go1.13.8
  major: "1"
  minor: "17"
  platform: darwin/amd64
serverVersion:
  buildDate: "2020-05-29T15:41:17Z"
  compiler: gc
  gitCommit: b20d34dbc46f336ed52ea4c01356da9952f0b8fc
  gitTreeState: clean
  gitVersion: v1.15.11-gke.17
  goVersion: go1.12.17b4
  major: "1"
  minor: 15+
  platform: linux/amd64
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by alexec** [comment]
on v2.8.0:

```
	if node.Outputs.Result != nil {
			// Support the case where item may be a map
			var itemMap map[string]wfv1.ItemValue
			err := json.Unmarshal([]byte(*node.Outputs.Result), &itemMap) // PANIC HERE
			if err == nil {
				resultsList = append(resultsList, wfv1.Item{Type: wfv1.Map, MapVal: itemMap})
			} else {
				resultsList = append(resultsList, wfv1.Item{Type: wfv1.String, StrVal: *node.Outputs.Result})
			}
```

**Comment by alexec** [comment]
I think `itemMap` is always nil on this line? So this line would always panic?!?!

**Comment by alexec** [comment]
@duongnt can you please attach the full YAML for the workflow that caused this? I.e. `argo get your-wf -o yaml`

**Comment by sarabala1979** [comment]
@alexec  I think `node.Outputs.Result` is nil that's why it is causing the panic. itemMap is ok

```
    jsonStr := `{"bala":"bala"}`
	pointer :=&jsonStr
	var itemMap map[string]string
	json.Unmarshal([]byte(*pointer), &itemMap)
	fmt.Println(len(itemMap))
	pointer = nil
	json.Unmarshal([]byte(*pointer), &itemMap). //Panic here
	fmt.Println(len(itemMap))
```

**Comment by sarabala1979** [comment]
Oh just saw that we have a check `if node.Outputs.Result != nil`

**Comment by duongnt** [comment]
> on v2.8.0:
> 
> ```
> 	if node.Outputs.Result != nil {
> 			// Support the case where item may be a map
> 			var itemMap map[string]wfv1.ItemValue
> 			err := json.Unmarshal([]byte(*node.Outputs.Result), &itemMap) // PANIC HERE
> 			if err == nil {
> 				resultsList = append(resultsList, wfv1.Item{Type: wfv1.Map, MapVal: itemMap})
> 			} else {
> 				resultsList = append(resultsList, wfv1.Item{Type: wfv1.String, StrVal: *node.Outputs.Result})
> 			}
> ```

@alexec Just to clarify, I didn't see this in 2.8.0. Only saw it when I upgraded to stable.

**Comment by alexec** [comment]
I'm wrong. The stack trace cannot be from v2.9.3 (stable). Line 2072 only contains a "}", that cannot panic. Can you please double check the controller logs for the version you are running?

![Screenshot 2020-07-17 at 10 41 23](https://user-images.githubusercontent.com/1142830/87815220-137ebf80-c81a-11ea-987a-de022f5680b3.png)

Looking at the code on v2.9.3 the only nil deref possible in `addOutputsToLocalScope` is on line 2070:

```
	for _, param := range outputs.Parameters {
		key := fmt.Sprintf("%s.outputs.parameters.%s", prefix, param.Name)
		if scope != nil {
			scope.addParamToScope(key, *param.Value)
		}
	}
```

This func has not changed recently.

**Comment by alexec** [comment]
@duongnt are any of your output parameters containing `valueFrom`?

**Comment by alexec** [comment]
Without further information or a way to reproduce this I don't have a way forward. I believe I know what line of code causes the panic, but I don't know what the underlying cause is. It is not happening to us, so I do believe it is something specific to @duongnt 's workflows.

**Comment by alexec** [comment]
Available for testing in [v2.10.0-rc1](https://github.com/argoproj/argo/releases/tag/v2.11.0-rc1).
